probe-rs/src/debug/variable.rs

```diff
@@ -1,41 +1,362 @@
+        }
+
+        // Is this an *add* or *update* operation?
+        let stored_key = if variable_to_add.variable_key == 0 {
+            // The caller is telling us this is definitely a new `Variable`
+            let new_cache_key: i64 = self.variable_cache_key.get() + 1;
+            variable_to_add.variable_key = new_cache_key;
+            match self
+                .variable_hash_map
+                .borrow_mut()
+                .insert(variable_to_add.variable_key, variable_to_add)
+            {
+                Some(old_variable) => {
+                    return Err(anyhow!("Attempt to insert a new `Variable`:{:?} with a duplicate cache key: {}. Please report this as a bug.", cache_variable.name, old_variable.variable_key).into());
+                }
+                None => {
+                    self.variable_cache_key.set(new_cache_key);
+                }
+            }
+            self.variable_cache_key.get()
+        } else {
+            // Attempt to update an existing `Variable` in the cache
+            let reused_cache_key = variable_to_add.variable_key;
+            if self
+                .variable_hash_map
+                .borrow_mut()
+                .insert(variable_to_add.variable_key, variable_to_add)
+                .is_none()
+            {
+                return Err(anyhow!("Attempt to update and existing `Variable`:{:?} with a non-existent cache key: {}. Please report this as a bug.", cache_variable.name, reused_cache_key).into());
+            }
+            reused_cache_key
+        };
+        // As the final act, we need to update the variable with an appropriate value.
+        // This requires distinct steps to ensure we don't get `borrow` conflicts on the variable cache.
+        if let Some(mut stored_variable) = self.get_variable_by_key(stored_key) {
+            stored_variable.extract_value(core, self);
+            if self
+                .variable_hash_map
+                .borrow_mut()
+                .insert(stored_variable.variable_key, stored_variable.clone())
+                .is_none()
+            {
+                Err(anyhow!("Failed to store variable at variable_cache_key: {}. Please report this as a bug.", stored_key).into())
+            } else {
+                Ok(stored_variable)
+            }
+        } else {
+            Err(anyhow!(
+                "Failed to store variable at variable_cache_key: {}. Please report this as a bug.",
+                stored_key
+            )
+            .into())
+        }
+    }
+
+    /// Retrieve a clone of a specific `Variable`, using the `variable_key`.
+    pub fn get_variable_by_key(&self, variable_key: i64) -> Option<Variable> {
+        self.variable_hash_map.borrow().get(&variable_key).cloned()
+    }
+
+    /// Retrieve a clone of a specific `Variable`, using the `name` and `parent_key`.
+    /// If there is more than one, it will be logged (log::warn!), and only the last will be returned.
```

Comment:
Good catch. Fixed now. Also, I changed the comment and the code to be a log::error() rather than a log::warn(). In theory, this should never happen, but if it does, it needs to be investigated. Thanks for catching :)